### PR TITLE
Feature/api crypto post purchase [ 5 ]

### PR DESCRIPTION
**Pull request by prakashchoudhary07**
Merge after #211

### CODE REVIEW

File: controllers/productsController.js [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+    return res.boom.badImplementation(ERROR_MESSAGE)
+  }
+}
+
+/**
+ * Make purchase request
+ * @param {Object} req - Express request object
+ * @param {Object} res - Express response object
+ */
+const makeTransaction = async (req, res) => {
```
Comment (Kratika0907) [inline controllers/productsController.js:74]: Please change the name as it is ambiguous. Try something like makePurchase

File: controllers/productsController.js [hunk 2]
```diff
@@ -0,0 +1,96 @@
...
+  try {
+    const { amount, items, totalQuantity } = req.body
+    const { id: userId } = req.userData
+    const purchaseResponse = await purchaseTransaction({ userId, amount, items, totalQuantity })
+    if (purchaseResponse) {
+      return res.json({
+        message: 'Transaction Successful.'
```
Comment (Kratika0907) [inline controllers/productsController.js:81]: Same for message. Purchase successful

File: models/products.js [hunk 3]
```diff
@@ -0,0 +1,107 @@
...
+    const isTransactionCompleted = await firestore.runTransaction(async t => {
+      const userWalletDocsArray = await t.get(userWalletRef)
+      const [userWalletDoc] = userWalletDocsArray.docs
+      const userWallet = await userWalletDoc.data()
+      if (checkSufficientAmountAvaliable(amount, userWallet.currencies)) {
+        const currencies = debitCoins(amount, userWallet.currencies)
+        const userWalletUpdateRef = walletsCollection.doc(userWalletDoc.id)
+        t.update(userWalletUpdateRef, { currencies })
+        const transactionDocument = transactionCollection.doc()
+        t.set(transactionDocument, {
```
Comment (Kratika0907) [inline models/products.js:84]: Please refer to data model for transaction collection

File: utils/crypto/transaction.js [hunk 4]
```diff
@@ -0,0 +1,45 @@
...
+      const debitCurrency = debit[key]
+      const availableCurrency = available[key]
+      if (debitCurrency <= availableCurrency) {
+        currenciesAvailableStatus.push(true)
+      } else {
+        currenciesAvailableStatus.push(false)
+      }
+    }
+  }
+  const status = !(currenciesAvailableStatus.some((val) => val === false))
```
Comment (Kratika0907) [inline utils/crypto/transaction.js:20]: Can we use simple Boolean filter here

Comment (Kratika0907) [reply utils/crypto/transaction.js:20]: Or we can have a key named  someTrue or someFalse rather than doing traversal of whole array

**Review by Kratika0907 — CHANGES_REQUESTED** [review]

File: models/crypto.js [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+    logger.error('Error retrieving product data', err)
+    throw err
+  }
+}
+
+/**
+ * @param Object : object with userId, amount, items, totalQuantity
+ * @return {Promise<boolean>}
+ */
+const purchaseTransaction = async ({ userId, amount, items, totalQuantity = null }) => {
```
[the remark this answers is not in the document]
Comment (prakashchoudhary07) [reply models/crypto.js]: Already update it in the previous commit

File: controllers/cryptoController.js [hunk 6]
```diff
@@ -0,0 +1,96 @@
...
+    return res.boom.badImplementation(ERROR_MESSAGE)
+  }
+}
+
+/**
+ * Make purchase request
+ * @param {Object} req - Express request object
+ * @param {Object} res - Express response object
+ */
+const makeTransaction = async (req, res) => {
```
[the remark this answers is not in the document]
Comment (prakashchoudhary07) [reply controllers/cryptoController.js]: Already update the name in the previous commit :smiley:

File: utils/crypto/transaction.js [hunk 7]
```diff
@@ -0,0 +1,34 @@
+/**
+ * Check if balance is available to make successful transaction
+ * @param {Object} debit - amount to be deduced (coins object)
+ * @param {Object} available - actual amount avaliable (conins object)
+ */
+function checkSufficientAmountAvaliable (debit, available) {
+  const brassAvailablity = (debit.brass <= available.brass)
```
[the remark this answers is not in the document]
Comment (prakashchoudhary07) [reply utils/crypto/transaction.js]: Made it generalized in the previous commit

File: models/products.js [hunk 8]
```diff
@@ -84,9 +84,11 @@ const purchaseTransaction = async ({ userId, amount, items, totalQuantity = null
         t.set(transactionDocument, {
           userId,
           amount,
-          items,
+          orders: items,
+          typeOfTransaction: 'products',
           totalQuantity,
-          timestamp: Date.now()
+          transactionId: transactionDocument.id,
+          'Date&Time': Date()
```
Comment (Kratika0907) [inline models/products.js:91]: I think we can change back to timeStamp here

File: utils/crypto/transaction.js [hunk 9]
```diff
@@ -5,20 +5,19 @@
...
     if (Object.hasOwnProperty.call(debit, key) && Object.hasOwnProperty.call(available, key)) {
       const debitCurrency = debit[key]
       const availableCurrency = available[key]
-      if (debitCurrency <= availableCurrency) {
-        currenciesAvailableStatus.push(true)
-      } else {
-        currenciesAvailableStatus.push(false)
+      if (!(debitCurrency <= availableCurrency)) {
+        isCurrenciesSufficient = false
       }
```
Comment (Kratika0907) [inline utils/crypto/transaction.js:15]: In both cases, we want to set it false?

**Review by Kratika0907 — CHANGES_REQUESTED** [review]
Requested one small change. Do let me know in case I missed something.

File: utils/crypto/transaction.js (hunk 9 above)
Comment (prakashchoudhary07) [reply utils/crypto/transaction.js:15]: If request currency and available currency don't match then if should return false

File: controllers/productsController.js [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+ * @param {Object} res - Express response object
+ */
+const addNewProduct = async (req, res) => {
+  try {
+    const productData = req.body
+    const product = await addProduct(productData)
+    if (product) {
+      return res.status(201).json({
+        message: 'Product added successfully!',
+        product
```
Comment (swarajpure) [inline controllers/productsController.js:36]: Do we need to return the newly created product?

File: controllers/productsController.js [hunk 11]
```diff
@@ -0,0 +1,96 @@
...
+ */
+const makePurchase = async (req, res) => {
+  try {
+    const { amount, items, totalQuantity } = req.body
+    const { id: userId } = req.userData
+    const purchaseResponse = await purchaseTransaction({ userId, amount, items, totalQuantity })
+    if (purchaseResponse) {
+      return res.json({
+        message: 'Purchase successful.'
+      })
```
Comment (swarajpure) [inline controllers/productsController.js:82]: Shouldn't this be 204?

File: routes/products.js [hunk 12]
```diff
@@ -0,0 +1,130 @@
...
+ *         content:
+ *           application/json:
+ *             schema:
+ *               $ref: '#/components/schemas/errors/badImplementation'
+ */
+router.post('/', authenticate, productsValidator.createProduct, productsController.addNewProduct)
+
+/**
+ * @swagger
+ * /products/{productid}:
```
Comment (swarajpure) [inline routes/products.js:72]: NIT: `/products/:productid`

File: test/integration/productsController.test.js [hunk 13]
```diff
@@ -0,0 +1,366 @@
...
+  let createdUserId
+  before(async function () {
+    const user = {
+      first_name: 'Prakash',
+      last_name: 'C',
+      yoe: 0,
+      img: './img.png',
+      github_id: 'prakashchoudhary07',
+      username: 'prakash'
+    }
```
Comment (swarajpure) [inline test/integration/productsController.test.js:26]: Better to have it coming from fixtures?

File: test/integration/productsController.test.js [hunk 14]
```diff
@@ -0,0 +1,366 @@
...
+    afterEach(async function () {
+      const product = {
+        id: 'coffee',
+        image: 'coffeeImage',
+        name: 'coffee',
+        manufacturer: 'RDS',
+        price: 55,
+        category: 'food',
+        usage: ['To make coffee']
+      }
```
Comment (swarajpure) [inline test/integration/productsController.test.js:43]: re: fixtures

File: test/integration/productsController.test.js [hunk 15]
```diff
@@ -0,0 +1,366 @@
...
+          expect(res.body.message).to.equal('Products returned successfully!')
+          expect(res.body.products).to.be.a('object')
+          expect(res.body.products.coffee).to.have.all.keys(
+            'id',
+            'image',
+            'name',
+            'manufacturer',
+            'price',
+            'category',
+            'usage'
```
Comment (swarajpure) [inline test/integration/productsController.test.js:82]: Better to have a list of these keys and using the spread operator here to check for the keys?

File: test/integration/productsController.test.js [hunk 16]
```diff
@@ -0,0 +1,366 @@
...
+  describe('GET /products/:productId', function () {
+    afterEach(async function () {
+      const product = {
+        id: 'milk',
+        image: 'milkImage',
+        name: 'milk',
+        manufacturer: 'RDS',
+        price: 55,
+        category: 'food',
+        usage: ['To make coffee']
```
Comment (swarajpure) [inline test/integration/productsController.test.js:97]: re: fixtures

File: test/integration/productsController.test.js [hunk 17]
```diff
@@ -0,0 +1,366 @@
...
+          expect(res.body.product).to.be.a('object')
+          expect(res.body.product).to.have.all.keys(
+            'id',
+            'image',
+            'name',
+            'manufacturer',
+            'price',
+            'category',
+            'usage'
+          )
```
Comment (swarajpure) [inline test/integration/productsController.test.js:138]: re: list

File: test/integration/productsController.test.js [hunk 18]
```diff
@@ -0,0 +1,366 @@
...
+        .post('/products')
+        .send({
+          id: 'water',
+          image: 'waterImage',
+          name: 'water',
+          manufacturer: 'RDS',
+          price: 55,
+          category: 'food',
+          usage: ['To drink ']
+        })
```
Comment (swarajpure) [inline test/integration/productsController.test.js:156]: re: fixtures

File: test/integration/productsController.test.js [hunk 19]
```diff
@@ -0,0 +1,366 @@
...
+        .request(app)
+        .post('/products')
+        .send({
+          image: 'coffeeImage',
+          name: 'coffee',
+          manufacturer: 'RDS',
+          price: 55,
+          category: 'food',
+          usage: ['To make coffee']
+        })
```
Comment (swarajpure) [inline test/integration/productsController.test.js:178]: re: fixtures

File: test/integration/productsController.test.js [hunk 20]
```diff
@@ -0,0 +1,366 @@
...
+        .post('/products')
+        .send({
+          id: 'water',
+          image: 'waterImage',
+          name: 'water',
+          manufacturer: 'RDS',
+          price: 55,
+          category: 'food',
+          usage: ['To drink ']
+        })
```
Comment (swarajpure) [inline test/integration/productsController.test.js:203]: re: fixtures

File: test/integration/productsController.test.js [hunk 21]
```diff
@@ -0,0 +1,366 @@
...
+          expect(res.body.product).to.be.a('object')
+          expect(res.body.product).to.have.all.keys(
+            'id',
+            'image',
+            'name',
+            'manufacturer',
+            'price',
+            'category',
+            'usage'
+          )
```
Comment (swarajpure) [inline test/integration/productsController.test.js:222]: re: list

**Review by swarajpure — CHANGES_REQUESTED** [review]
Have some suggestions, please check 🙂

File: controllers/productsController.js (hunk 11 above)
Comment (prakashchoudhary07) [reply controllers/productsController.js:82]: Nope, 204 means no content, but here we are returning message :sweat_smile:

File: controllers/productsController.js (hunk 10 above)
Comment (prakashchoudhary07) [reply controllers/productsController.js:36]: Not necessary, But this was returned as it was mentioned in the API contract as far as I remember

File: controllers/products.js [hunk 22]
```diff
@@ -0,0 +1,96 @@
...
+  try {
+    const { amount, items, totalQuantity } = req.body
+    const { id: userId } = req.userData
+    const purchaseResponse = await purchaseTransaction({ userId, amount, items, totalQuantity })
+    if (purchaseResponse) {
+      return res.json({
+        message: 'Purchase successful.'
+      })
+    }
+    return res.boom.paymentRequired('Insufficient coins.')
```
Comment (swarajpure) [inline controllers/products.js:84]: Shouldn't this be 403?

File: controllers/products.js [hunk 23]
```diff
@@ -0,0 +1,96 @@
...
+ * @param {Object} res - Express response object
+ */
+const makePurchase = async (req, res) => {
+  try {
+    const { amount, items, totalQuantity } = req.body
+    const { id: userId } = req.userData
+    const purchaseResponse = await purchaseTransaction({ userId, amount, items, totalQuantity })
+    if (purchaseResponse) {
+      return res.json({
+        message: 'Purchase successful.'
```
Comment (swarajpure) [inline controllers/products.js:81]: NIT: `!` instead of `.`?

File: controllers/products.js [hunk 24]
```diff
@@ -0,0 +1,96 @@
...
+ * @param {Object} res - Express response object
+ */
+
+const getProduct = async (req, res) => {
+  try {
+    const productId = req.params.productId
+    const productData = await fetchProduct(productId)
+    if (productData) {
+      return res.json({
+        message: 'Product returned successfully.',
```
Comment (swarajpure) [inline controllers/products.js:58]: NIT: `!` instead of `.`?

File: middlewares/validators/crypto-products.js [hunk 25]
```diff
@@ -0,0 +1,46 @@
...
+  }
+}
+
+const purchaseTransaction = async (req, res, next) => {
+  const schema = joi.object().keys({
+    amount: joi.object().keys().required(),
+    items: joi.array().items({
+      itemId: joi.string().required(),
+      name: joi.string().optional(),
+      quantity: joi.number()
```
Comment (swarajpure) [inline middlewares/validators/crypto-products.js:29]: `quantity` optional or required?

File: models/products.js [hunk 26]
```diff
@@ -0,0 +1,109 @@
+const firestore = require('../utils/firestore')
+const { checkSufficientAmountAvaliable, debitCoins } = require('../utils/crypto/transaction')
+const productsCollection = firestore.collection('crypto-products')
+const walletsCollection = firestore.collection('wallets') // Users collection name to be standardized
+const transactionCollection = firestore.collection('transactions') // Users collection name to be standardized
```
Comment (swarajpure) [inline models/products.js:5]: Can we have them coming from constants?

**Review by swarajpure — CHANGES_REQUESTED** [review]
Some doubts and some suggestions, please check

File: middlewares/validators/crypto-products.js (hunk 25 above)
Comment (prakashchoudhary07) [reply middlewares/validators/crypto-products.js]: Its required

**Comment by prakashchoudhary07** [comment]
This is being deleted as the contracts were updated, and the project is on hold currently

**Pull request closed without merging**